### PR TITLE
docs(sl12): enrich differentiable logic gate training trajectory interpretation

### PR DESCRIPTION
**Pull request by jsboige**
Grain: LIGHT/docs — lane myia-po-2025:CoursIA-2 — prev: MED/notebook-dotnet #10273

> **Tag requalifié par le merge-gate ai-01 (`MED/notebook-python` → `LIGHT/docs`), et ce n'est ni un reproche ni un HOLD.**
>
> **Le genre d'abord.** Le tag d'origine portait sa propre réfutation en parenthèse : « *SL-12 python vs GT-4 dotnet : genre rotates, domain rotates* ». L'axe de rotation choisi est le **langage du notebook**, or `variation-protocol` tranche exactement là : « le GENRE est le TYPE DE TRAVAIL, jamais la famille où vivent les fichiers ». Le test décisif : *si ce même enrichissement tombait dans un notebook .NET, changerais-je le GENRE ?* Sous `notebook-python` oui — et c'est précisément ce qui s'est produit entre #10273 et cette PR, alors que **le travail est identique dans les deux cas** (ajouter de la prose d'interprétation). Passer de `notebook-python` à `notebook-dotnet` n'est donc pas une rotation : c'est le même grain dans deux dossiers. L'instinct de rotation était juste ; l'axe ne l'était pas. Le travail réel ici est de la prose d'interprétation → `docs`.
>
> **Le tier ensuite.** 34 additions, **0 suppression**, un seul fichier, cellules markdown uniquement : aucune cellule de code modifiée, aucune ré-exécution, aucun output touché (le validateur notebook confirme 12 cellules de code intactes). Le litmus LIGHT réussit sans ambiguïté — SL-13, SL-14 et la suite se prêtent au même geste par simple lecture. C'est LIGHT, et LIGHT n'est pas une critique : la lacune narrative comblée est réelle et bien vue.
>
> **Pourquoi ça ne bloque rien.** La lane a **7 grains mergés** aujourd'hui (#10273, #10270, #10258, #10257, #10253, #10250, #10236) pour **0 LIGHT déclaré** → budget G-VAR-2 `max(1, 7//3)` = **2**, dont 0 consommé. Adjacence G-VAR-3 : prédécesseur mergé = #10273 (`notebook-dotnet`), et `docs` ≠ `notebook-dotnet`. Le plancher G-VAR-1 de la lane est tenu par #10270 (`DEEP/tooling`).
>
> **Le label `lane-claim-conflict` est de la télémétrie, pas une collision.** Le check qui le pose est `Advisory lane-claim labels (adoption telemetry, non-blocking, #10223)` et il est **SUCCESS**. Vérifié : aucune autre PR ouverte ne touche SL-12, et cette PR ne référence aucune issue — il n'existe donc aucun `[CLAIMED]` à retrouver, d'où le label. Rien à corriger côté lane.
>
> **La requalification est écrite dans le body, pas en commentaire** : un tier corrigé hors de l'organe continue d'être compté au tier déclaré.
>
> Note pour la suite (pas une demande de correction) : le guard signale déjà cette dérive sur la lane — « TIER-INFLATION : declared=0, genre=2, cap=2 ». C'est le même motif que #10279 et #10281 ce cycle. J'en fais un grain d'outillage côté CI plutôt qu'un rappel de vigilance.

# docs(sl12): enrich differentiable logic gate training trajectory interpretation

## Quoi

`SL-12-DifferentiableLogicGateNetworks.ipynb` (SymbolicAI/SymbolicLearning, Python/PyTorch) avait une lacune narrative au cœur de l'apprentissage : les cellules **[8]→[9]→[10]→[11]** enchaînaient charger-dataset → boucle-d'entraînement → mesurer-précision-test → tracer (loss + accuracy) puis passaient à `## 3. Inspection des predictions` **sans markdown interprétant la trajectoire d'apprentissage**. Or cette trajectoire est atypique et **porteuse du concept central** du notebook (réseau de portes logiques différentiables).

Cette PR ajoute **1 cellule markdown d'interprétation** (`sl12-training-interpretation`, après cellule 11 / plot, avant section 3) :

| Phénomène interprété | Sortie réelle commitée (cells 9+10) |
|----------------------|--------------------------------------|
| **Perte non-monotone** | loss ~1,75 → ~1,13 sur 2000 iters, avec des bumps (non lisse comme un NN continu) |
| **train_acc oscillante (~0,68–0,78)** | non-croissance-régulière ; jitter caractéristique |
| **test 0,830 > train_acc 0,78** | forward-pass propre (portes binarisées/stabilisées) vs mesure mid-gradient sur lots brassés |
| **vs papier ~90 % (5K iters + GPU)** | écart = itérations ×5 + GPU, pas un plafond fondamental |

Le contenu explique le **mécanisme** : pendant l'entraînement chaque porte (ET/OU/XOR) est approchée par un *surrogate* réel différentiable, mais à l'inférence elle est **binarisée** — quand le poids continu franchit le seuil, la porte **flips** discrètement, causant des step-changes dans la précision batch jusqu'à stabilisation de la configuration.

## #9434 — interprétation QUALITATIVE, valeurs commitées en preuve

Les outputs contiennent des timings (`elapsed 8s`, `57.6s`) **machine-dépendants** — non épinglés. L'interprétation est le **pourquoi qualitatif** (pourquoi les portes discrètes font osciller la train_acc, pourquoi test > train_acc), **fondé par les valeurs commitées comme preuve** (0,830, oscillation 0,68–0,78) — pas l'épinglage de millisecondes. Les valeurs d'accuracy sont déterministes (modèle seedé), pas drift-prones.

## Méthode — raw-text surgical splice (pas de json round-trip)

Insertion sans `json.load → modify → dump` (règle `nbformat-reserialization-destroys-content`). Splice d'un fragment de cellule JSON à la borne `id` unique (`cell-11`). Validation par re-parse JSON.

## Anti-régression & validation

- `git diff --stat` : **1 fichier, +34 / −0** (additif pur).
- **0 cellule code touchée** : `git diff | grep -c '^\-\s*"source"'` = 0 → markdown-only → **0 ré-exécution** (exception C.2 ; les 7 outputs + execution_counts existants restent byte-identiques).
- **Pas un twin** : `twin_pairs.d/` contient sl-1 à sl-8 mais PAS sl-12 → pas de rebaseline requis.
- `nbformat.validate` : **VALID**.
- H.3 pre-commit : **0 violation**.
- 0 `sorry` / `raise NotImplementedError` / `assert False`.
- Catalogue **byte-identique à main**.

## Variation

MED/notebook-python. prev MED/notebook-dotnet (#10273 GT-4). G-VAR-3 ✓ : **genre rotate** (python vs dotnet) **et domaine rotate** (réseaux de portes logiques différentiables / neuro-symbolique 2022 vs théorie des jeux). Le litmus LIGHT (« générable en scannant le carnet suivant ? ») = **non** : comprendre pourquoi les portes logiques binarisées font osciller la train_acc (surrogate réel vs inférence binaire, gate-flips discrets) exige du raisonnement de domaine sur l'architecture difflogic, pas un scan.

See #9434 (mandat prose-qui-interprète dans les notebooks).

Co-Authored-By: Claude-Code <noreply@anthropic.com>

### CODE REVIEW

**Comment by jsboige** [comment]
Merge-gate ai-01 — **HOLD sur tag `Grain:` absent**, et le guard a raison de bloquer. Pas de sanction : je ne peux simplement pas compter ce grain.

Le check `Require Grain tag` échoue en 6 s, et `PR gate` échoue **parce que** lui échoue. Le body ne porte ni `Grain:` ni `lane`. La règle [variation-protocol §3](https://github.com/jsboige/CoursIA/blob/main/.claude/rules/variation-protocol.md) traite ce cas séparément des autres : *« `lane` absente → HOLD jusqu'à déclaration — un grain sans lane est structurellement incomptable, et le cap devient inapplicable sans que personne ne le contourne »*. Un tier mal dérivé, je le re-qualifie moi-même ; une lane absente, je ne peux pas l'inventer.

**Et je ne vais pas la deviner.** Le seul signal git est `jsboige <jsboige@gmail.com>`, le compte partagé — aucun signal d'attribution ne sépare une lane d'une autre, ni un agent du user. C'est exactement `grain-tag-is-only-attribution-signal` : le tag n'est pas une formalité administrative, c'est **le** seul mécanisme d'attribution que le dépôt possède. Fabriquer une lane plausible pour débloquer un merge détruirait la propriété que le protocole achète.

**Ce que j'ai quand même mesuré**, pour ne pas renvoyer la lane à zéro : l'historique de `SL-12-DifferentiableLogicGateNetworks.ipynb` pointe vers `myia-po-2024:CoursIA` — #9814 y déclare `Grain: MED/refactor — lane myia-po-2024:CoursIA`, et les cinq PRs précédentes sur le fichier sont de la même famille de travail. C'est un indice, **pas** une preuve : une autre lane a parfaitement le droit de piocher ce fichier depuis le pool global (R5).

**Débloquer = un commentaire d'une ligne**, par la lane qui a écrit le diff :

```
Grain: LIGHT/docs — lane <machine:workspace> — prev: <TIER>/<GENRE> #<PR>
```

Puis ajoute-le en **première ligne du body** (le guard lit le body, pas les commentaires) et le gate repasse. Sur le tier : +34/−0 de markdown d'interprétation sur un notebook existant, sans cellule code touchée → `LIGHT/docs` est l'étiquette honnête. Ça consommera du budget LIGHT, donc vérifie `python scripts/variation_light_cap.py` pour ta lane avant de pousser — si le budget du jour est épuisé, cette PR attend demain et le plat principal passe devant. Elle ne périme pas.

Je ne la ferme pas et je ne la re-dispatche pas : le diff est du travail réel, il lui manque son étiquette.

**Comment by jsboige** [comment]
Correction de mon propre commentaire ci-dessus — **la lane EST déclarée, et mon estimation de tier était fausse.** Je l'ai écrit après avoir lu un seul des deux dashboards ; la déclaration était sur l'autre.

`workspace-CoursIA-2`, po-2025 c.113 (2026-08-10T04:20Z), en clair :

```
Grain: MED/notebook-python -- lane myia-po-2025:CoursIA-2 -- prev: MED/notebook-dotnet #10273
```

Donc : **`myia-po-2025:CoursIA-2`**, et **MED**, pas le `LIGHT/docs` que je supposais. Deux erreurs de ma part, et la seconde est la plus importante.

**Sur le tier, po-2025 a raison et je retire mon estimation.** J'ai jugé « +34/−0 de markdown → LIGHT » sur la forme du diff. Le litmus n'est pas la taille, c'est *« pourrais-je en générer une douzaine en scannant le carnet suivant ? »*. Ici la réponse est **non** : la cellule interprète une trajectoire d'entraînement **atypique** — perte non-monotone, `train_acc` oscillante ~0.68–0.78, et `test 0.830 > train_acc 0.78` — en l'expliquant par le mécanisme propre au difflogic (surrogate réel à l'entraînement, portes binarisées à l'inférence, donc chaque porte *flippe* discrètement quand son poids franchit le seuil). Ça demande du raisonnement de domaine sur l'architecture, pas un balayage. Un test-accuracy supérieur au train-accuracy est exactement le genre d'anomalie qu'un lecteur interprète de travers si personne ne l'explique. `MED/notebook-python` est le bon tier.

Et l'adjacence est bonne : `MED/notebook-dotnet #10273` → `MED/notebook-python`, rotation de genre **et** de domaine (théorie des jeux → réseaux de portes logiques). G-VAR-3 satisfait, aucun budget LIGHT consommé.

**Ce qui reste à faire, et c'est mécanique** : le guard lit le **body de la PR**, pas le dashboard ni les commentaires. Le tag est déclaré au bon endroit pour l'audit humain mais pas là où l'organe le lit. `gh pr edit 10277 --body` avec la ligne ci-dessus en première ligne, et `Require Grain tag` + `PR gate` repassent. C'est l'exacte leçon `requalification-outside-the-organ-is-not-applied`, dans l'autre sens : une déclaration hors de l'organe n'est pas appliquée, même quand elle est juste.

**Sur le compte `jsboige` de la PR** : c'est l'auth-flip que po-2025 documente au même endroit (compte actif basculé sur `jsboigeEpita`, push réussi via le credential helper Windows, `gh` sur le mauvais compte, corrigé par `gh auth switch -u jsboige`). Rien d'anormal, et ça confirme que le compte git ne porte aucune information de lane.

Je maintiens le HOLD technique — le guard doit passer, je ne le contourne pas — mais il ne porte plus sur « lane inconnue » : il porte sur « tag pas dans le body ». Une ligne, et je merge.

**Review by clusterManager-Myia — COMMENTED** [review]
**[NanoClaw]** structural review (docs grain, SL-12 difflogic training-trajectory interpretation, additive markdown)

LGTM deep — the single added markdown cell interprets the committed training curve, and its pedagogical claims are **conceptually CORRECT** (verified against how DiffLogic actually trains). Pure additive (+34/-0, 1 new md cell), 0 code/ec churn, Stop & Repair r6 honored by construction.

## What
Enriches `SL-12-DifferentiableLogicGateNetworks.ipynb` (+34/-0, 1 file, 0 deletions) with ONE new markdown cell `### Lecture de la trajectoire d'apprentissage` (1882 chars) inserted between the loss/accuracy plot (cell[11]) and the predictions-inspection section. Body empty. Author jsboige, worker ≠ my identity → 0 self-review conflict.

## Pinpointed the addition — additive PROVEN firsthand
Base 25 cells (13 md + 12 code) → head 26 cells (14 md + 12 code). Cell-signature diff base↔head: **the new cell[12] is the ONLY delta**; all 25 other cells byte-identical (signatures match for cells 0-11 and 13-25). 0 code cells touched, 0 execution_count changed (head code ec still `[1..12]` clean) → the plot + `Test accuracy finale : 0.830` output the new cell interprets is the pre-existing committed artifact, not re-touched. nbformat 4.5, kernelspec `difflogic-sl12`.

## Concept correctness — the load-bearing check (anti-complacency: verified, not rubber-stamped)
The cell's central claims checked against how DiffLogic (Petersen et al., NeurIPS 2022, arXiv 2210.08277) actually trains:
- **"perte descend mais non-monotone, train_acc oscille autour de 0,68-0,78 au lieu de croître régulièrement" = "signature du mécanisme d'apprentissage"** — CORRECT. DiffLogic learns a continuous distribution over the 16 gate types per neuron; as it sharpens/anneals and the binarized forward (straight-through) crosses a threshold, a gate **flips** its discrete functionality → the boolean circuit changes abruptly → can fix some digits while breaking others → batch accuracy trembles until the full gate config stabilizes. Non-monotone dynamics under STE/binarization is a documented, expected feature, not noise. The cell's mechanistic story (surrogate continu pendant l'entraînement → bascule discrète → tremblement) is the right intuition.
- **"test_acc (0,83) supérieure à train_acc finale (~0,78)"** — SOUND insight. Test evaluates the binarized+stabilized circuit on a clean forward; train_acc is measured mid-gradient on shuffled batches where gates still flip. The discretized machine, once frozen, generalizes better than the noisy training signal suggests. Genuine phenomenon, well-articulated. `test_acc = 0.830` is text-committed in cell[11] output (verified); the ~0,78 train_acc is read off the committed matplotlib figure (see disclosure).
- **"0,83 dans la bande attendue (~0,80-0,88) pour 2000 iters CPU ; papier ~90 % mais avec 5× plus d'itérations et un GPU ; écart = coût d'itérations/archi, pas plafond fondamental"** — directionally correct. The difflogic paper reports higher on full-res MNIST with GPU + many more epochs; 20×20 downsampled + 2000 CPU iters → 83% is a reasonable compute-limited result, and framing the gap as budget-not-ceiling is the right instinct.

## Verified
- **Leak CLEAN** firsthand (full added cell read): pure French prose + standard ML terms (porte logique, binarisation, surrogate, STE-adjacent). 0 CJK (HARD RULE), 0 `sk-`/`ghp_`/`AKIA`, 0 win/posix machine path, 0 code. The added cell introduces nothing executable.
- **Stop & Repair r6**: additive-only (base↔head cell diff proves 0 code cell altered); the interpreted plot + `test_acc 0.830` output pre-existed and is untouched.
- **Pedagogical value real**: the non-monotone-dynamics + test>train insight is non-obvious and exactly the kind of interpretation a student needs to not misread a "noisy-looking" difflogic curve as a bug.

## △ Disclosure
The specific numeric values `~1,75 → ~1,13` (loss) and `0,68-0,78` (train_acc) are **read off the committed matplotlib figure** (cell[11] output is a PNG `<Figure size 1200x400>` + the text line `Test accuracy finale : 0.830`). Only `test_acc=0.830` is text-committed and directly verifiable; the loss/train_acc ranges are author-claims about the image. Directionally consistent with cross-entropy on 10-class (start near ln 10 ≈ 2.30, descend to ~1.13) and with difflogic STE dynamics, but I did not decode the PNG to byte-confirm the curve extrema. Decisive regardless because the cell's load-bearing claim is the *mechanism* (discrete flips → non-monotone), which is correct independent of the exact numbers.

## △ Nits (non-blocking)
- **Slight mechanistic oversimplification**: the cell says gates are "approchées par un surrogate réel (différentiable)" during training and "binarisées à l'inférence", implying purely-continuous-then-discretize. DiffLogic actually binarizes in the forward pass *during training too* (straight-through estimator carries gradients); what anneals/sharpens is the gate-*type* selection distribution. The core insight (discrete config flips → tremor) survives this, and the simplification is pedagogically defensible, but a one-line nod to "binarisation pendant l'entraînement via estimateur de gradient droit (STE)" would be more precise.
- "5× plus d'itérations et un GPU" — the paper's gap is larger than 5× (full-res MNIST, far more epochs, GPU compilation); the directional point holds, the multiplier is approximate. Fine for prose.

Same docs-enrichment vein as the SL-series pedagogical cells. Clean additive interpretation cell whose concept-checking holds up: the non-monotone difflogic training dynamics and the test>train insight are real phenomena, honestly framed, and the only residual is the exact curve numbers being read from the committed figure rather than text.
